[JENKINS-60641, JENKINS-60644] incorrectly converted AntClassLoader.loadResource to a Java 5 for-loop

jglick opened the pull request:
[JENKINS-60644](https://issues.jenkins-ci.org/browse/JENKINS-60644)

Corrects a mistake in #4254 which broke some plugins.

- [X] fix confirmed via ATH
- [x] fix confirmed here via JTH

### Proposed changelog entries

* Corrects a critical regression in Jenkins 2.212 affecting at least the `script-security` and `active-directory` plugins.

jglick committed:
#4254 incorrectly converted AntClassLoader.loadResource to a Java 5 for-loop.

oleg-nenashev reviewed (approved):
Thanks for catching! I was unable to notice it in code reviews even after pin-pointing the pull request. Will need an out=of order release if @kohsuke is available

zbynek commented:
@jglick I guess the same fix is needed in the new `getUrl` method which also changed from iterating a vector to an arrayList

jglick commented:
> I guess the same fix is needed in the new `getUrl` method

No, that one looks fine to me, since it is checking for a null result in each iteration. The problem here in `loadResource` was that it was looping over all path components but actually ignoring everything but the last one.

jglick committed:
Reproduced error in functional test with a small sample plugin.
Without fix, Jenkins startup fails:
java.lang.NullPointerException
	at java.io.Reader.<init>(Reader.java:78)
	at java.io.InputStreamReader.<init>(InputStreamReader.java:113)
	at org.apache.commons.io.IOUtils.copy(IOUtils.java:2440)
	at org.apache.commons.io.IOUtils.toString(IOUtils.java:1084)
	at io.jenkins.plugins.loads_resource.Main.stuff(Main.java:36)

oleg-nenashev commented on test/src/test/java/jenkins/util/AntClassLoaderTest.java:
```diff
@@ -0,0 +1,47 @@
+    @WithPlugin("loads-resource.jpi")
+    @Test
+    public void loadsResource() throws Exception {
+        assertNotNull(r.jenkins.pluginManager.getPlugin("loads-resource").classLoader.getResourceAsStream("io/jenkins/plugins/loads_resource/stuff"));
```
This is probably the only case where `PluginManager.getPlugin("loads-resource")` actually works in JTH :(

jglick replied:
Well, it would always work for loaded plugins, but normally the `classLoader` would just be `AppClassLoader`. You need to go out of your way to test the real plugin loaders.

jglick commented on core/src/main/java/jenkins/util/AntClassLoader.java:
```diff
@@ -744,12 +744,13 @@ public InputStream getResourceAsStream(String name) {
+            }
         }
-        return stream;
+        return null;
```
For those who like this sort of thing:

```java
return pathComponents.stream().map(pc -> getResourceStream(pc, name)).filter(Objects::nonNull).findFirst().orElse(null);
```

oleg-nenashev replied:
I don't
